### PR TITLE
feat: Allow users to configure the internal thread pool (#11)

### PR DESCRIPTION
**Pull request by alhridoy**
This PR introduces a new feature that allows users to configure the internal thread pool used in the `_raw_indexing_method` function of the `EarthEngineBackendArray` class. 

A new optional parameter `executor_kwargs` has been added to the function. This parameter is a dictionary that can be used to pass configuration options to the `ThreadPoolExecutor`. If `executor_kwargs` is not provided, an empty dictionary is used, preserving the current behavior.

This feature provides users with more flexibility and control over the execution of the function, especially in scenarios where specific thread pool configurations are needed.

Changes:
- Added `executor_kwargs` parameter to `_raw_indexing_method` function in `xee/ext.py`.

Please review and let me know if any changes are required.

### CODE REVIEW

File: configurable-thread-pool [hunk 1]
```diff
@@ -0,0 +1 @@
+Branch 'configurable-thread-pool' set up to track remote branch 'main' from 'origin'.
```
Comment (alxmrs) [inline configurable-thread-pool:1]: Please remove this file.

File: xee/ext.py [hunk 2]
```diff
@@ -634,7 +634,7 @@ def reduce_bands(x, acc):
     return target_image
 
   def _raw_indexing_method(
-      self, key: tuple[Union[int, slice], ...]
+      self, key: tuple[Union[int, slice], ...], executor_kwargs: Optional[dict] = None
```
Comment (alxmrs) [inline xee/ext.py:637]: This extra argument is named well, but I think it should be set in the constructor.

File: xee/ext.py [hunk 3]
```diff
@@ -682,8 +682,11 @@ def _raw_indexing_method(
         for _ in range(shape[0])
     ]
 
-    # TODO(#11): Allow users to configure this via kwargs.
-    with concurrent.futures.ThreadPoolExecutor() as pool:
+     # If executor_kwargs is None, use an empty dictionary
+    if executor_kwargs is None:
+      executor_kwargs = {}
```
Comment (alxmrs) [inline xee/ext.py:687]: Let's do this bit of logic in the constructor, too.

File: xee/ext.py [hunk 4]
```diff
@@ -682,8 +682,11 @@ def _raw_indexing_method(
...
+    # Pass executor_kwargs to ThreadPoolExecutor
```
Comment (alxmrs) [inline xee/ext.py:688]: Comment is a bit too verbose for my taste.

**Review by alxmrs — CHANGES_REQUESTED** [review]
Thanks for your contribution, @alhridoy! I have some initial feedback. 

A couple more high-level notes: 
- Make sure that users can set this from `open_dataset()`. To accomplish this, the new keyword arg needs to be accepted in `EarthEngineBackendEntrypoint()`.
- Please update our integration tests to prove that everything is working.

**Comment by alhridoy** [comment]
Hi @alxmrs when i run the test file, this error pops up.

`Could not import runpy module
Traceback (most recent call last):
  File "<frozen runpy>", line 15, in <module>
  File "<frozen importlib.util>", line 16, in <module>
  File "/Users/hridoy/Desktop/open_source/Xee/xee/types.py", line 16, in <module>
    from typing import Union, TypedDict
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/typing.py", line 27, in <module>
    import contextlib
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 7, in <module>
    from functools import wraps
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/functools.py", line 22, in <module>
    from types import GenericAlias
ImportError: cannot import name 'GenericAlias' from partially initialized module 'types' (most likely due to a circular import) (/Users/hridoy/Desktop/open_source/Xee/xee/types.py)` 
I renamed the types.py to type.py then I run the test again, it runs but no output is shown in the terminal. any guidance to run the ext_integration_test.py file?

**Comment by alhridoy** [comment]
@alxmrs Thanks for your guidance. Would you check if these changes are working properly? if not,  Please let me necessary the changes I need to make. Thanks

File: xee/ext.py [hunk 5]
```diff
@@ -89,6 +89,12 @@ class EarthEngineStore(common.AbstractDataStore):
 
   DEFAULT_MASK_VALUE = np.iinfo(np.int32).max
 
+  def __init__(self, executor_kwargs=None):
```
Comment (alxmrs) [inline xee/ext.py:92]: There already is a constructor.

File: xee/ext_integration_test.py [hunk 6]
```diff
@@ -358,6 +358,18 @@ def test_data_sanity_check(self):
     self.assertNotEqual(temperature_2m.min(), 0.0)
     self.assertNotEqual(temperature_2m.max(), 0.0)
 
+  def test_open_dataset_with_executor_kwargs(self):
```
Comment (alxmrs) [inline xee/ext_integration_test.py:361]: Thanks for the test!

**Comment by alxmrs** [comment]
Please fix the merge conflicts in this branch.

**Comment by alhridoy** [comment]
Hi @alxmrs . Thanks for the feedback. ,I added  a new parameter executor_kwargs has been added to the __init__ method also resolved merge conflict.

File: xee/ext.py [hunk 7]
```diff
@@ -203,6 +204,13 @@ def __init__(
         coordinates=f'{self.primary_dim_name} {x_dim_name} {y_dim_name}',
         crs=self.crs_arg,
     )
+
+    # Initialize executor_kwargs
+    if executor_kwargs is None:
+      self.executor_kwargs = {}
+    else:
+      self.executor_kwargs = executor_kwargs
```
Comment (alxmrs) [inline xee/ext.py:212]:
```suggestion
    if executor_kwargs is None:
      executor_kwargs = {}
    self.executor_kwargs = executor_kwargs
```

File: xee/ext.py [hunk 8]
```diff
@@ -706,7 +714,7 @@ def reduce_bands(x, acc):
     return target_image
 
   def _raw_indexing_method(
-      self, key: tuple[Union[int, slice], ...]
+      self, key: tuple[Union[int, slice], ...]
```
Comment (alxmrs) [inline xee/ext.py:717]: Please remove extra space (to minimize the diff).

File: xee/ext.py (hunk 7 above)
Comment (alxmrs) [reply xee/ext.py:212]: Please initialize this at the top or bottom of the `__init__`.

File: xee/ext.py [hunk 9]
```diff
@@ -845,6 +854,7 @@ def open_dataset(
       primary_dim_name: Optional[str] = None,
       primary_dim_property: Optional[str] = None,
       ee_mask_value: Optional[float] = None,
+      executor_kwargs: Optional[dict] = None,
```
Comment (alxmrs) [inline xee/ext.py:857]: In addition to adding it to the arguments, please update the docstring.

**Review by alxmrs — COMMENTED** [review]
Definitely looking better. Here are a few more additional suggestions.

**Comment by alhridoy** [comment]
Hi @alxmrs , Would you check the new commits address your comments? If there is still any issue  I would be happy to work on that.

**Comment by alxmrs** [comment]
It feels like you are using a bot to make this patch. I'll be happy to work with you to add this feature. First, please do a self review before I make any additional reviews.

**Comment by alhridoy** [comment]
Thanks! Please give me couple of weeks.

**Pull request closed without merging**